internal/fs/parallel_dirops_test.go

```diff
@@ -0,0 +1,687 @@
+				"file2.txt":              "xyz",
+				"explicitDir1/":          "",
+				"explicitDir1/file1.txt": "12345",
+				"explicitDir1/file2.txt": "6789101112",
+				"implicitDir1/file1.txt": "-1234556789",
+			}))
+}
+
+func (t *ParallelDiropsTest) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		FileSystemConfig: config.FileSystemConfig{
+			DisableParallelDirops: false,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *ParallelDiropsWithoutCachesTest) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		FileSystemConfig: config.FileSystemConfig{
+			DisableParallelDirops: false,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.serverCfg.DirTypeCacheTTL = 0
+	t.serverCfg.InodeAttributeCacheTTL = 0
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *ParallelDiropsTest) SetupTest() {
+	t.createFilesAndDirStructureInBucket()
+}
+
+func (t *ParallelDiropsTest) TearDownTest() {
+	t.fsTest.TearDown()
+}
+
+func (t *ParallelDiropsTest) TearDownSuite() {
+	t.fsTest.TearDownTestSuite()
+}
+
+func (t *ParallelDiropsTest) TestParallelLookUpsForSameFile() {
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(filePath)
+		return fileInfo, err
+	}
+	var stat1, stat2 os.FileInfo
+	var err1, err2 error
+
+	// Parallel lookups of file just under mount.
+	filePath := path.Join(mntDir, "file1.txt")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.Equal(t.T(), int64(6), stat1.Size())
+	assert.Equal(t.T(), int64(6), stat2.Size())
+	assert.Contains(t.T(), filePath, stat1.Name())
+	assert.Contains(t.T(), filePath, stat2.Name())
+
+	// Parallel lookups of file under a directory in mount.
+	filePath = path.Join(mntDir, "explicitDir1/file2.txt")
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.Equal(t.T(), int64(10), stat1.Size())
+	assert.Equal(t.T(), int64(10), stat2.Size())
+	assert.Contains(t.T(), filePath, stat1.Name())
+	assert.Contains(t.T(), filePath, stat2.Name())
+}
+
+func (t *ParallelDiropsTest) TestParallelLookUpsForSameDir() {
+	lookUpFunc := func(wg *sync.WaitGroup, dirPath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(dirPath)
+		return fileInfo, err
+	}
+	var stat1, stat2 os.FileInfo
+	var err1, err2 error
+
+	// Parallel lookups of explicit dir under mount.
+	dirPath := path.Join(mntDir, "explicitDir1")
```

Comment:
Similarly seems replication for explicitDir1, we can write a common method which will be used by both implicitdir and explicit dir.

Reply:
Resolving. see the reasoning: https://github.com/GoogleCloudPlatform/gcsfuse/pull/1906#discussion_r1606294076